http/client.go

```diff
@@ -14,30 +14,45 @@ import (
 	"time"
 )
 
-// New creates a new http.Client with a custom certificate, which can be loaded from the passed CA Bundle file and/or
+type TransportOption func(*http.Transport) error
+
+// New creates a new http.Client with the Passed Transport Options
+// that will have a custom certificate if it is loaded from the passed CA Bundle file and/or
 // directory. If both CABundleFile and CABundleDir are empty arguments, it creates an unsecure HTTP client.
-func New(CABundleFile, CABundleDir string, httpTimeout time.Duration) (*http.Client, error) {
-	return _new(CABundleFile, CABundleDir, httpTimeout, "")
-}
+func New(httpTimeout time.Duration, opts ...TransportOption) (*http.Client, error) {
+	t := &http.Transport{
+		TLSClientConfig: &tls.Config{},
+	}
+	for _, opt := range opts {
+		if err := opt(t); err != nil {
+			return nil, err
+		}
+	}
 
-// NewAcceptInvalidHostname new http.Client with ability to accept HTTPS certificates that don't
-// match the hostname of the server they are connecting to.
-func NewAcceptInvalidHostname(CABundleFile, CABundleDir string, httpTimeout time.Duration, hostname string) (*http.Client, error) {
-	return _new(CABundleFile, CABundleDir, httpTimeout, hostname)
+	return &http.Client{
+		Timeout:   httpTimeout,
+		Transport: t,
+	}, nil
 }
 
-func _new(CABundleFile, CABundleDir string, httpTimeout time.Duration, acceptInvalidHostname string) (*http.Client, error) {
-	// go default http transport settings
-	t := &http.Transport{}
-
-	if CABundleFile != "" || CABundleDir != "" {
-		certs, err := getCertPool(CABundleFile, CABundleDir)
-		if err != nil {
-			return nil, err
+func WithCABundle(CABundleFile, CABundleDir string) TransportOption {
+	return func(t *http.Transport) error {
+		if CABundleFile != "" || CABundleDir != "" {
+			certs, err := getCertPool(CABundleFile, CABundleDir)
+			if err != nil {
+				return err
+			}
+			t.TLSClientConfig.RootCAs = certs
 		}
+		return nil
+	}
+}
 
-		t.TLSClientConfig = &tls.Config{
-			RootCAs: certs,
+func WithAcceptInvalidHostname(CABundleFile, CABundleDir, acceptInvalidHostname string) TransportOption {
```

Comment:
I'm not 100% sure how `acceptInvalidHostname` relates to `CABundleFile` and `CABundleDir`. Is it a requirement for them to be tied together?

Reply:
I think it is, because if you check the VerifyOptions definition on the VerifyPeerCertificate, defines the Roots as the t.TLSClientConfig.RootCAs that are assigned on WithCABundle

Reply:
If I got this correctly (which I'm not very sure about), this is a very convoluted way to convince the cert validation to check against a different hostname, instead of the one actually used for resolving the address.

It might be used together with custom CA bundles, but I don't see why it should be coupled. Do you have a different understanding?